From c55772bb32b5d5f5c27daa02d4dfbbb099c8b523 Mon Sep 17 00:00:00 2001
From: Hao Chen <chenhao@loongson.cn>
Date: Thu, 30 May 2024 19:26:19 +0800
Subject: [PATCH 03/15] FROMLIST: LoongArch64: Optimize functions in loopfilter
 module

Optimize functions in loopfilter module with lsx and lasx.

Link: https://bitbucket.org/multicoreware/x265_git/pull-requests/31
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 source/CMakeLists.txt                        |    2 +
 source/common/CMakeLists.txt                 |   19 +
 source/common/loongarch64/asm-primitives.cpp |   71 +
 source/common/loongarch64/loongson_asm.S     |  769 ++++++++++
 source/common/loongarch64/loopfilter.S       | 1324 ++++++++++++++++++
 source/common/loongarch64/loopfilter.h       |   57 +
 source/test/testharness.h                    |    5 +
 7 files changed, 2247 insertions(+)
 create mode 100644 source/common/loongarch64/asm-primitives.cpp
 create mode 100644 source/common/loongarch64/loongson_asm.S
 create mode 100644 source/common/loongarch64/loopfilter.S
 create mode 100644 source/common/loongarch64/loopfilter.h

diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index 785015ec1..bb5ed3fbc 100755
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -410,6 +410,8 @@ elseif(NASM_FOUND AND X86)
         message(STATUS "Found Nasm ${NASM_VERSION_STRING} to build assembly primitives")
         option(ENABLE_ASSEMBLY "Enable use of assembly coded primitives" ON)
     endif()
+elseif(LOONGARCH64)
+    option(ENABLE_ASSEMBLY "Enable use of assembly coded primitives" ON)
 else()
     option(ENABLE_ASSEMBLY "Enable use of assembly coded primitives" OFF)
 endif()
diff --git a/source/common/CMakeLists.txt b/source/common/CMakeLists.txt
index ec526593b..b18482349 100644
--- a/source/common/CMakeLists.txt
+++ b/source/common/CMakeLists.txt
@@ -133,6 +133,25 @@ if(POWER)
     endif()
 endif()
 
+if(ENABLE_ASSEMBLY AND LOONGARCH64)
+    if(GCC AND (CMAKE_CXX_FLAGS_RELEASE MATCHES "-O3"))
+        message(STATUS "Detected CXX compiler using -O3 optimization level")
+    endif()
+
+    set(C_SRCS asm-primitives.cpp loopfilter.h)
+    enable_language(ASM)
+
+    # add loongarch assembly/intrinsic files here
+    set(A_SRCS loopfilter.S)
+
+    set(LOONGARCH64_ASMS "${A_SRCS}" CACHE INTERNAL "LOONGARCH64 Assembly Sources")
+    foreach(SRC ${C_SRCS})
+        set(ASM_PRIMITIVES ${ASM_PRIMITIVES} loongarch64/${SRC})
+    endforeach()
+    source_group(Assembly FILES ${ASM_PRIMITIVES})
+
+endif(ENABLE_ASSEMBLY AND LOONGARCH64)
+
 # set_target_properties can't do list expansion
 string(REPLACE ";" " " VERSION_FLAGS "${VFLAGS}")
 set_source_files_properties(version.cpp PROPERTIES COMPILE_FLAGS ${VERSION_FLAGS})
diff --git a/source/common/loongarch64/asm-primitives.cpp b/source/common/loongarch64/asm-primitives.cpp
new file mode 100644
index 000000000..53ba228a7
--- /dev/null
+++ b/source/common/loongarch64/asm-primitives.cpp
@@ -0,0 +1,71 @@
+/*****************************************************************************
+ * Copyright (C) 2013-2024 MulticoreWare, Inc
+ *
+ * Authors: Hao Chen <chenhao@loongson.cn>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02111, USA.
+ *
+ * This program is also available under a commercial proprietary license.
+ * For more information, contact us at license @ x265.com.
+ *****************************************************************************/
+
+#include "common.h"
+#include "primitives.h"
+#include "x265.h"
+#include "cpu.h"
+#include "loopfilter.h"
+
+namespace X265_NS {
+// private x265 namespace
+
+
+void setupAssemblyPrimitives(EncoderPrimitives &p, int cpuMask)
+{
+#if HAVE_LSX
+    if (cpuMask & X265_CPU_LSX)
+    {
+        // loopfilter
+        p.saoCuOrgE0 = PFX(saoCuOrgE0_lsx);
+        p.saoCuOrgE1 = PFX(saoCuOrgE1_lsx);
+        p.saoCuOrgE1_2Rows = PFX(saoCuOrgE1_2Rows_lsx);
+        p.saoCuOrgE2[0] = PFX(saoCuOrgE2_lsx);
+        p.saoCuOrgE2[1] = PFX(saoCuOrgE2_lsx);
+        p.saoCuOrgE3[0] = PFX(saoCuOrgE3_lsx);
+        p.saoCuOrgE3[1] = PFX(saoCuOrgE3_lsx);
+        p.saoCuOrgB0 = PFX(saoCuOrgB0_lsx);
+        p.sign = PFX(calSign_lsx);
+        p.pelFilterLumaStrong[0] = PFX(pelFilterLumaStrong_V_lsx);
+        p.pelFilterLumaStrong[1] = PFX(pelFilterLumaStrong_H_lsx);
+        p.pelFilterChroma[0] = PFX(pelFilterChroma_V_lsx);
+        p.pelFilterChroma[1] = PFX(pelFilterChroma_H_lsx);
+        p.saoCuStatsE0 = PFX(saoCuStatsE0_lsx);
+        p.saoCuStatsE1 = PFX(saoCuStatsE1_lsx);
+        p.saoCuStatsE2 = PFX(saoCuStatsE2_lsx);
+        p.saoCuStatsE3 = PFX(saoCuStatsE3_lsx);
+        p.saoCuStatsBO = PFX(saoCuStatsB0_lsx);
+
+    }
+#endif /* HAVE_LSX */
+
+#if HAVE_LASX
+    if (cpuMask & X265_CPU_LASX)
+    {
+        // loopfilter
+        p.sign = PFX(calSign_lasx);
+    }
+#endif /* HAVE_LASX */
+}
+
+}
diff --git a/source/common/loongarch64/loongson_asm.S b/source/common/loongarch64/loongson_asm.S
new file mode 100644
index 000000000..c00082ab1
--- /dev/null
+++ b/source/common/loongarch64/loongson_asm.S
@@ -0,0 +1,769 @@
+/*
+*****************************************************************************
+* loongson_asm.S: loongson asm abstraction layer
+*****************************************************************************
+* Copyright (c) 2024 Loongson Technology Corporation Limited
+* Copyright (C) 2013-2024 MulticoreWarae, Inc
+*
+* Authors: gxw <guxiwei-hf@loongson.cn>
+*          Shiyou Yin<yinshiyou-hf@loongson.cn>
+*
+* Permission to use, copy, modify, and/or distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+* WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+* MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+* ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+* WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+* ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+* OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*****************************************************************************
+*/
+
+/*
+ * This file is a LoongArch assembly helper file and available under ISC
+ * license. It provides a large number of macros and alias to simplify
+ * writing assembly code, especially for LSX and LASX optimizations.
+ *
+ * Any one can modify it or add new features for his/her own purposes.
+ * Contributing a patch will be appreciated as it might be useful for
+ * others as well. Send patches to loongson contributor mentioned above.
+ *
+ * MAJOR version: Usage changes, incompatible with previous version.
+ * MINOR version: Add new macros/functions, or bug fixes.
+ * MICRO version: Comment changes or implementation changes.
+ */
+
+#define LML_VERSION_MAJOR 0
+#define LML_VERSION_MINOR 4
+#define LML_VERSION_MICRO 0
+
+#define ASM_PREF
+#define DEFAULT_ALIGN    5
+
+.macro function name, align=DEFAULT_ALIGN
+.macro endfunc
+    jirl    $r0, $r1, 0x0
+    .size ASM_PREF\name, . - ASM_PREF\name
+    .purgem endfunc
+.endm
+.text ;
+.align \align ;
+.globl ASM_PREF\name ;
+.type  ASM_PREF\name, @function ;
+ASM_PREF\name: ;
+.endm
+
+.macro  const name, align=DEFAULT_ALIGN
+    .macro endconst
+    .size  \name, . - \name
+    .purgem endconst
+    .endm
+.section .rodata
+.align   \align
+\name:
+.endm
+
+/*
+ *============================================================================
+ * LoongArch register alias
+ *============================================================================
+ */
+
+#define a0 $a0
+#define a1 $a1
+#define a2 $a2
+#define a3 $a3
+#define a4 $a4
+#define a5 $a5
+#define a6 $a6
+#define a7 $a7
+
+#define t0 $t0
+#define t1 $t1
+#define t2 $t2
+#define t3 $t3
+#define t4 $t4
+#define t5 $t5
+#define t6 $t6
+#define t7 $t7
+#define t8 $t8
+
+#define s0 $s0
+#define s1 $s1
+#define s2 $s2
+#define s3 $s3
+#define s4 $s4
+#define s5 $s5
+#define s6 $s6
+#define s7 $s7
+#define s8 $s8
+
+#define zero $zero
+#define sp   $sp
+#define ra   $ra
+
+#define fa0  $fa0
+#define fa1  $fa1
+#define fa2  $fa2
+#define fa3  $fa3
+#define fa4  $fa4
+#define fa5  $fa5
+#define fa6  $fa6
+#define fa7  $fa7
+#define ft0  $ft0
+#define ft1  $ft1
+#define ft2  $ft2
+#define ft3  $ft3
+#define ft4  $ft4
+#define ft5  $ft5
+#define ft6  $ft6
+#define ft7  $ft7
+#define ft8  $ft8
+#define ft9  $ft9
+#define ft10 $ft10
+#define ft11 $ft11
+#define ft12 $ft12
+#define ft13 $ft13
+#define ft14 $ft14
+#define ft15 $ft15
+#define fs0  $fs0
+#define fs1  $fs1
+#define fs2  $fs2
+#define fs3  $fs3
+#define fs4  $fs4
+#define fs5  $fs5
+#define fs6  $fs6
+#define fs7  $fs7
+
+#define f0  $f0
+#define f1  $f1
+#define f2  $f2
+#define f3  $f3
+#define f4  $f4
+#define f5  $f5
+#define f6  $f6
+#define f7  $f7
+#define f8  $f8
+#define f9  $f9
+#define f10 $f10
+#define f11 $f11
+#define f12 $f12
+#define f13 $f13
+#define f14 $f14
+#define f15 $f15
+#define f16 $f16
+#define f17 $f17
+#define f18 $f18
+#define f19 $f19
+#define f20 $f20
+#define f21 $f21
+#define f22 $f22
+#define f23 $f23
+#define f24 $f24
+#define f25 $f25
+#define f26 $f26
+#define f27 $f27
+#define f28 $f28
+#define f29 $f29
+#define f30 $f30
+#define f31 $f31
+
+#define vr0 $vr0
+#define vr1 $vr1
+#define vr2 $vr2
+#define vr3 $vr3
+#define vr4 $vr4
+#define vr5 $vr5
+#define vr6 $vr6
+#define vr7 $vr7
+#define vr8 $vr8
+#define vr9 $vr9
+#define vr10 $vr10
+#define vr11 $vr11
+#define vr12 $vr12
+#define vr13 $vr13
+#define vr14 $vr14
+#define vr15 $vr15
+#define vr16 $vr16
+#define vr17 $vr17
+#define vr18 $vr18
+#define vr19 $vr19
+#define vr20 $vr20
+#define vr21 $vr21
+#define vr22 $vr22
+#define vr23 $vr23
+#define vr24 $vr24
+#define vr25 $vr25
+#define vr26 $vr26
+#define vr27 $vr27
+#define vr28 $vr28
+#define vr29 $vr29
+#define vr30 $vr30
+#define vr31 $vr31
+
+#define xr0 $xr0
+#define xr1 $xr1
+#define xr2 $xr2
+#define xr3 $xr3
+#define xr4 $xr4
+#define xr5 $xr5
+#define xr6 $xr6
+#define xr7 $xr7
+#define xr8 $xr8
+#define xr9 $xr9
+#define xr10 $xr10
+#define xr11 $xr11
+#define xr12 $xr12
+#define xr13 $xr13
+#define xr14 $xr14
+#define xr15 $xr15
+#define xr16 $xr16
+#define xr17 $xr17
+#define xr18 $xr18
+#define xr19 $xr19
+#define xr20 $xr20
+#define xr21 $xr21
+#define xr22 $xr22
+#define xr23 $xr23
+#define xr24 $xr24
+#define xr25 $xr25
+#define xr26 $xr26
+#define xr27 $xr27
+#define xr28 $xr28
+#define xr29 $xr29
+#define xr30 $xr30
+#define xr31 $xr31
+
+/*
+ *============================================================================
+ * LSX/LASX synthesize instructions
+ *============================================================================
+ */
+
+/*
+ * Description : Dot product of byte vector elements
+ * Arguments   : Inputs  - vj, vk
+ *               Outputs - vd
+ *               Return Type - halfword
+ */
+.macro vdp2.h.bu vd, vj, vk
+    vmulwev.h.bu      \vd,    \vj,    \vk
+    vmaddwod.h.bu     \vd,    \vj,    \vk
+.endm
+
+.macro vdp2.h.bu.b vd, vj, vk
+    vmulwev.h.bu.b    \vd,    \vj,    \vk
+    vmaddwod.h.bu.b   \vd,    \vj,    \vk
+.endm
+
+.macro vdp2.w.h vd, vj, vk
+    vmulwev.w.h       \vd,    \vj,    \vk
+    vmaddwod.w.h      \vd,    \vj,    \vk
+.endm
+
+.macro xvdp2.h.bu xd, xj, xk
+    xvmulwev.h.bu    \xd,    \xj,    \xk
+    xvmaddwod.h.bu   \xd,    \xj,    \xk
+.endm
+
+.macro xvdp2.h.bu.b xd, xj, xk
+    xvmulwev.h.bu.b    \xd,  \xj,    \xk
+    xvmaddwod.h.bu.b   \xd,  \xj,    \xk
+.endm
+
+.macro xvdp2.w.h xd, xj, xk
+    xvmulwev.w.h       \xd,  \xj,    \xk
+    xvmaddwod.w.h      \xd,  \xj,    \xk
+.endm
+
+/*
+ * Description : Dot product & addition of halfword vector elements
+ * Arguments   : Inputs  - vj, vk
+ *               Outputs - vd
+ *               Return Type - twice size of input
+ */
+.macro vdp2add.h.bu vd, vj, vk
+    vmaddwev.h.bu     \vd,    \vj,    \vk
+    vmaddwod.h.bu     \vd,    \vj,    \vk
+.endm
+
+.macro vdp2add.h.bu.b vd, vj, vk
+    vmaddwev.h.bu.b   \vd,    \vj,    \vk
+    vmaddwod.h.bu.b   \vd,    \vj,    \vk
+.endm
+
+.macro vdp2add.w.h vd, vj, vk
+    vmaddwev.w.h      \vd,    \vj,    \vk
+    vmaddwod.w.h      \vd,    \vj,    \vk
+.endm
+
+.macro xvdp2add.h.bu.b xd, xj, xk
+    xvmaddwev.h.bu.b   \xd,  \xj,    \xk
+    xvmaddwod.h.bu.b   \xd,  \xj,    \xk
+.endm
+
+.macro xvdp2add.w.h xd, xj, xk
+    xvmaddwev.w.h      \xd,  \xj,    \xk
+    xvmaddwod.w.h      \xd,  \xj,    \xk
+.endm
+
+/*
+ * Description : Range element vj[i] to vk[i] ~ vj[i]
+ * clip: vj > vk ? vj : vk && vj < va ? vj : va
+ */
+.macro vclip.h  vd,  vj, vk, va
+    vmax.h    \vd,  \vj,   \vk
+    vmin.h    \vd,  \vd,   \va
+.endm
+
+.macro vclip.w  vd,  vj, vk, va
+    vmax.w    \vd,  \vj,   \vk
+    vmin.w    \vd,  \vd,   \va
+.endm
+
+.macro xvclip.h  xd,  xj, xk, xa
+    xvmax.h    \xd,  \xj,   \xk
+    xvmin.h    \xd,  \xd,   \xa
+.endm
+
+.macro xvclip.w  xd,  xj, xk, xa
+    xvmax.w    \xd,  \xj,   \xk
+    xvmin.w    \xd,  \xd,   \xa
+.endm
+
+/*
+ * Description : Range element vj[i] to 0 ~ 255
+ * clip255: vj < 255 ? vj : 255 && vj > 0 ? vj : 0
+ */
+.macro vclip255.h  vd, vj
+    vmaxi.h   \vd,   \vj,  0
+    vsat.hu   \vd,   \vd,  7
+.endm
+
+.macro vclip255.w  vd, vj
+    vmaxi.w   \vd,   \vj,  0
+    vsat.wu   \vd,   \vd,  7
+.endm
+
+.macro xvclip255.h  xd, xj
+    xvmaxi.h   \xd,   \xj,  0
+    xvsat.hu   \xd,   \xd,  7
+.endm
+
+.macro xvclip255.w  xd, xj
+    xvmaxi.w   \xd,   \xj,  0
+    xvsat.wu   \xd,   \xd,  7
+.endm
+
+/*
+ * Description : Store elements of vector
+ * vd : Data vector to be stroed
+ * rk : Address of data storage
+ * ra : Offset of address
+ * si : Index of data in vd
+ */
+.macro vstelmx.b vd, rk, ra, si
+    add.d      \rk,  \rk,  \ra
+    vstelm.b   \vd,  \rk,  0, \si
+.endm
+
+.macro vstelmx.h vd, rk, ra, si
+    add.d      \rk,  \rk,  \ra
+    vstelm.h   \vd,  \rk,  0, \si
+.endm
+
+.macro vstelmx.w vd, rk, ra, si
+    add.d      \rk,  \rk,  \ra
+    vstelm.w   \vd,  \rk,  0, \si
+.endm
+
+.macro vstelmx.d  vd, rk, ra, si
+    add.d      \rk,  \rk,  \ra
+    vstelm.d   \vd,  \rk,  0, \si
+.endm
+
+.macro vmov xd, xj
+    vor.v  \xd,  \xj,  \xj
+.endm
+
+.macro xmov xd, xj
+    xvor.v  \xd,  \xj,  \xj
+.endm
+
+.macro xvstelmx.d  xd, rk, ra, si
+    add.d      \rk, \rk,  \ra
+    xvstelm.d  \xd, \rk,  0, \si
+.endm
+
+/*
+ *============================================================================
+ * LSX/LASX custom macros
+ *============================================================================
+ */
+
+/*
+ * Load 4 float, double, V128, v256 elements with stride.
+ */
+.macro FLDS_LOADX_4 src, stride, stride2, stride3, out0, out1, out2, out3
+    fld.s     \out0,    \src,    0
+    fldx.s    \out1,    \src,    \stride
+    fldx.s    \out2,    \src,    \stride2
+    fldx.s    \out3,    \src,    \stride3
+.endm
+
+.macro FLDD_LOADX_4 src, stride, stride2, stride3, out0, out1, out2, out3
+    fld.d     \out0,    \src,    0
+    fldx.d    \out1,    \src,    \stride
+    fldx.d    \out2,    \src,    \stride2
+    fldx.d    \out3,    \src,    \stride3
+.endm
+
+.macro LSX_LOADX_4 src, stride, stride2, stride3, out0, out1, out2, out3
+    vld     \out0,    \src,    0
+    vldx    \out1,    \src,    \stride
+    vldx    \out2,    \src,    \stride2
+    vldx    \out3,    \src,    \stride3
+.endm
+
+.macro LASX_LOADX_4 src, stride, stride2, stride3, out0, out1, out2, out3
+    xvld    \out0,    \src,    0
+    xvldx   \out1,    \src,    \stride
+    xvldx   \out2,    \src,    \stride2
+    xvldx   \out3,    \src,    \stride3
+.endm
+
+/*
+ * Description : Transpose 4x4 block with half-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ */
+.macro LSX_TRANSPOSE4x4_H in0, in1, in2, in3, out0, out1, out2, out3, \
+                          tmp0, tmp1
+    vilvl.h   \tmp0,  \in1,   \in0
+    vilvl.h   \tmp1,  \in3,   \in2
+    vilvl.w   \out0,  \tmp1,  \tmp0
+    vilvh.w   \out2,  \tmp1,  \tmp0
+    vilvh.d   \out1,  \out0,  \out0
+    vilvh.d   \out3,  \out0,  \out2
+.endm
+
+/*
+ * Description : Transpose 4x4 block with word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ * Details     :
+ * Example     :
+ *               1, 2, 3, 4            1, 5, 9,13
+ *               5, 6, 7, 8    to      2, 6,10,14
+ *               9,10,11,12  =====>    3, 7,11,15
+ *              13,14,15,16            4, 8,12,16
+ */
+.macro LSX_TRANSPOSE4x4_W in0, in1, in2, in3, out0, out1, out2, out3, \
+                          tmp0, tmp1
+
+    vilvl.w    \tmp0,   \in1,    \in0
+    vilvh.w    \out1,   \in1,    \in0
+    vilvl.w    \tmp1,   \in3,    \in2
+    vilvh.w    \out3,   \in3,    \in2
+
+    vilvl.d    \out0,   \tmp1,   \tmp0
+    vilvl.d    \out2,   \out3,   \out1
+    vilvh.d    \out3,   \out3,   \out1
+    vilvh.d    \out1,   \tmp1,   \tmp0
+.endm
+
+/*
+ * Description : Transpose 8x8 block with half-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3, in4, in5, in6, in7
+ *               Outputs - out0, out1, out2, out3, out4, out5, out6, out7
+ */
+.macro LSX_TRANSPOSE8x8_H in0, in1, in2, in3, in4, in5, in6, in7, out0, out1,   \
+                          out2, out3, out4, out5, out6, out7, tmp0, tmp1, tmp2, \
+                          tmp3, tmp4, tmp5, tmp6, tmp7
+    vilvl.h      \tmp0,    \in6,   \in4
+    vilvl.h      \tmp1,    \in7,   \in5
+    vilvl.h      \tmp2,    \in2,   \in0
+    vilvl.h      \tmp3,    \in3,   \in1
+
+    vilvl.h      \tmp4,    \tmp1,  \tmp0
+    vilvh.h      \tmp5,    \tmp1,  \tmp0
+    vilvl.h      \tmp6,    \tmp3,  \tmp2
+    vilvh.h      \tmp7,    \tmp3,  \tmp2
+
+    vilvh.h      \tmp0,    \in6,   \in4
+    vilvh.h      \tmp1,    \in7,   \in5
+    vilvh.h      \tmp2,    \in2,   \in0
+    vilvh.h      \tmp3,    \in3,   \in1
+
+    vpickev.d    \out0,    \tmp4,  \tmp6
+    vpickod.d    \out1,    \tmp4,  \tmp6
+    vpickev.d    \out2,    \tmp5,  \tmp7
+    vpickod.d    \out3,    \tmp5,  \tmp7
+
+    vilvl.h      \tmp4,    \tmp1,  \tmp0
+    vilvh.h      \tmp5,    \tmp1,  \tmp0
+    vilvl.h      \tmp6,    \tmp3,  \tmp2
+    vilvh.h      \tmp7,    \tmp3,  \tmp2
+
+    vpickev.d    \out4,    \tmp4,  \tmp6
+    vpickod.d    \out5,    \tmp4,  \tmp6
+    vpickev.d    \out6,    \tmp5,  \tmp7
+    vpickod.d    \out7,    \tmp5,  \tmp7
+.endm
+
+/*
+ * Description : Transpose 16x8 block with byte elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3, in4, in5, in6, in7
+ *               Outputs - out0, out1, out2, out3, out4, out5, out6, out7
+ */
+.macro LASX_TRANSPOSE16X8_B in0, in1, in2, in3, in4, in5, in6, in7,        \
+                            in8, in9, in10, in11, in12, in13, in14, in15,  \
+                            out0, out1, out2, out3, out4, out5, out6, out7,\
+                            tmp0, tmp1, tmp2, tmp3, tmp4, tmp5, tmp6, tmp7
+    xvilvl.b   \tmp0,    \in2,     \in0
+    xvilvl.b   \tmp1,    \in3,     \in1
+    xvilvl.b   \tmp2,    \in6,     \in4
+    xvilvl.b   \tmp3,    \in7,     \in5
+    xvilvl.b   \tmp4,    \in10,    \in8
+    xvilvl.b   \tmp5,    \in11,    \in9
+    xvilvl.b   \tmp6,    \in14,    \in12
+    xvilvl.b   \tmp7,    \in15,    \in13
+    xvilvl.b   \out0,    \tmp1,    \tmp0
+    xvilvh.b   \out1,    \tmp1,    \tmp0
+    xvilvl.b   \out2,    \tmp3,    \tmp2
+    xvilvh.b   \out3,    \tmp3,    \tmp2
+    xvilvl.b   \out4,    \tmp5,    \tmp4
+    xvilvh.b   \out5,    \tmp5,    \tmp4
+    xvilvl.b   \out6,    \tmp7,    \tmp6
+    xvilvh.b   \out7,    \tmp7,    \tmp6
+    xvilvl.w   \tmp0,    \out2,    \out0
+    xvilvh.w   \tmp2,    \out2,    \out0
+    xvilvl.w   \tmp4,    \out3,    \out1
+    xvilvh.w   \tmp6,    \out3,    \out1
+    xvilvl.w   \tmp1,    \out6,    \out4
+    xvilvh.w   \tmp3,    \out6,    \out4
+    xvilvl.w   \tmp5,    \out7,    \out5
+    xvilvh.w   \tmp7,    \out7,    \out5
+    xvilvl.d   \out0,    \tmp1,    \tmp0
+    xvilvh.d   \out1,    \tmp1,    \tmp0
+    xvilvl.d   \out2,    \tmp3,    \tmp2
+    xvilvh.d   \out3,    \tmp3,    \tmp2
+    xvilvl.d   \out4,    \tmp5,    \tmp4
+    xvilvh.d   \out5,    \tmp5,    \tmp4
+    xvilvl.d   \out6,    \tmp7,    \tmp6
+    xvilvh.d   \out7,    \tmp7,    \tmp6
+.endm
+
+/*
+ * Description : Transpose 4x4 block with half-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ */
+.macro LASX_TRANSPOSE4x4_H in0, in1, in2, in3, out0, out1, out2, out3, \
+                           tmp0, tmp1
+    xvilvl.h   \tmp0,  \in1,   \in0
+    xvilvl.h   \tmp1,  \in3,   \in2
+    xvilvl.w   \out0,  \tmp1,  \tmp0
+    xvilvh.w   \out2,  \tmp1,  \tmp0
+    xvilvh.d   \out1,  \out0,  \out0
+    xvilvh.d   \out3,  \out0,  \out2
+.endm
+
+/*
+ * Description : Transpose 4x8 block with half-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ */
+.macro LASX_TRANSPOSE4x8_H in0, in1, in2, in3, out0, out1, out2, out3, \
+                           tmp0, tmp1
+    xvilvl.h      \tmp0,    \in2,   \in0
+    xvilvl.h      \tmp1,    \in3,   \in1
+    xvilvl.h      \out2,    \tmp1,  \tmp0
+    xvilvh.h      \out3,    \tmp1,  \tmp0
+
+    xvilvl.d      \out0,    \out2,  \out2
+    xvilvh.d      \out1,    \out2,  \out2
+    xvilvl.d      \out2,    \out3,  \out3
+    xvilvh.d      \out3,    \out3,  \out3
+.endm
+
+/*
+ * Description : Transpose 8x8 block with half-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3, in4, in5, in6, in7
+ *               Outputs - out0, out1, out2, out3, out4, out5, out6, out7
+ */
+.macro LASX_TRANSPOSE8x8_H in0, in1, in2, in3, in4, in5, in6, in7,         \
+                           out0, out1, out2, out3, out4, out5, out6, out7, \
+                           tmp0, tmp1, tmp2, tmp3, tmp4, tmp5, tmp6, tmp7
+    xvilvl.h     \tmp0,   \in6,     \in4
+    xvilvl.h     \tmp1,   \in7,     \in5
+    xvilvl.h     \tmp2,   \in2,     \in0
+    xvilvl.h     \tmp3,   \in3,     \in1
+
+    xvilvl.h     \tmp4,   \tmp1,    \tmp0
+    xvilvh.h     \tmp5,   \tmp1,    \tmp0
+    xvilvl.h     \tmp6,   \tmp3,    \tmp2
+    xvilvh.h     \tmp7,   \tmp3,    \tmp2
+
+    xvilvh.h     \tmp0,   \in6,     \in4
+    xvilvh.h     \tmp1,   \in7,     \in5
+    xvilvh.h     \tmp2,   \in2,     \in0
+    xvilvh.h     \tmp3,   \in3,     \in1
+
+    xvpickev.d   \out0,   \tmp4,    \tmp6
+    xvpickod.d   \out1,   \tmp4,    \tmp6
+    xvpickev.d   \out2,   \tmp5,    \tmp7
+    xvpickod.d   \out3,   \tmp5,    \tmp7
+
+    xvilvl.h     \tmp4,   \tmp1,    \tmp0
+    xvilvh.h     \tmp5,   \tmp1,    \tmp0
+    xvilvl.h     \tmp6,   \tmp3,    \tmp2
+    xvilvh.h     \tmp7,   \tmp3,    \tmp2
+
+    xvpickev.d   \out4,   \tmp4,    \tmp6
+    xvpickod.d   \out5,   \tmp4,    \tmp6
+    xvpickev.d   \out6,   \tmp5,    \tmp7
+    xvpickod.d   \out7,   \tmp5,    \tmp7
+.endm
+
+/*
+ * Description : Transpose 2x4x4 block with half-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ */
+.macro LASX_TRANSPOSE2x4x4_H in0, in1, in2, in3, out0, out1, out2, out3, \
+                             tmp0, tmp1, tmp2
+    xvilvh.h   \tmp1,    \in0,     \in1
+    xvilvl.h   \out1,    \in0,     \in1
+    xvilvh.h   \tmp0,    \in2,     \in3
+    xvilvl.h   \out3,    \in2,     \in3
+
+    xvilvh.w   \tmp2,    \out3,    \out1
+    xvilvl.w   \out3,    \out3,    \out1
+
+    xvilvl.w   \out2,    \tmp0,    \tmp1
+    xvilvh.w   \tmp1,    \tmp0,    \tmp1
+
+    xvilvh.d   \out0,    \out2,    \out3
+    xvilvl.d   \out2,    \out2,    \out3
+    xvilvh.d   \out1,    \tmp1,    \tmp2
+    xvilvl.d   \out3,    \tmp1,    \tmp2
+.endm
+
+/*
+ * Description : Transpose 4x4 block with word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ * Details     :
+ * Example     :
+ *               1, 2, 3, 4,  1, 2, 3, 4        1,5, 9,13, 1,5, 9,13
+ *               5, 6, 7, 8,  5, 6, 7, 8   to   2,6,10,14, 2,6,10,14
+ *               9,10,11,12,  9,10,11,12 =====> 3,7,11,15, 3,7,11,15
+ *              13,14,15,16, 13,14,15,16        4,8,12,16, 4,8,12,16
+ */
+.macro LASX_TRANSPOSE4x4_W in0, in1, in2, in3, out0, out1, out2, out3, \
+                           tmp0, tmp1
+
+    xvilvl.w    \tmp0,   \in1,    \in0
+    xvilvh.w    \out1,   \in1,    \in0
+    xvilvl.w    \tmp1,   \in3,    \in2
+    xvilvh.w    \out3,   \in3,    \in2
+
+    xvilvl.d    \out0,   \tmp1,   \tmp0
+    xvilvl.d    \out2,   \out3,   \out1
+    xvilvh.d    \out3,   \out3,   \out1
+    xvilvh.d    \out1,   \tmp1,   \tmp0
+.endm
+
+/*
+ * Description : Transpose 8x8 block with word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3, in4, in5, in6, in7
+ *               Outputs - out0, out1, out2, out3, out4, out5, out6,
+ *               _out7
+ * Example     : LASX_TRANSPOSE8x8_W
+ *         in0 : 1,2,3,4,5,6,7,8
+ *         in1 : 2,2,3,4,5,6,7,8
+ *         in2 : 3,2,3,4,5,6,7,8
+ *         in3 : 4,2,3,4,5,6,7,8
+ *         in4 : 5,2,3,4,5,6,7,8
+ *         in5 : 6,2,3,4,5,6,7,8
+ *         in6 : 7,2,3,4,5,6,7,8
+ *         in7 : 8,2,3,4,5,6,7,8
+ *
+ *        out0 : 1,2,3,4,5,6,7,8
+ *        out1 : 2,2,2,2,2,2,2,2
+ *        out2 : 3,3,3,3,3,3,3,3
+ *        out3 : 4,4,4,4,4,4,4,4
+ *        out4 : 5,5,5,5,5,5,5,5
+ *        out5 : 6,6,6,6,6,6,6,6
+ *        out6 : 7,7,7,7,7,7,7,7
+ *        out7 : 8,8,8,8,8,8,8,8
+ */
+.macro LASX_TRANSPOSE8x8_W in0, in1, in2, in3, in4, in5, in6, in7,\
+                           out0, out1, out2, out3, out4, out5, out6, out7,\
+                           tmp0, tmp1, tmp2, tmp3
+    xvilvl.w    \tmp0,   \in2,    \in0
+    xvilvl.w    \tmp1,   \in3,    \in1
+    xvilvh.w    \tmp2,   \in2,    \in0
+    xvilvh.w    \tmp3,   \in3,    \in1
+    xvilvl.w    \out0,   \tmp1,   \tmp0
+    xvilvh.w    \out1,   \tmp1,   \tmp0
+    xvilvl.w    \out2,   \tmp3,   \tmp2
+    xvilvh.w    \out3,   \tmp3,   \tmp2
+
+    xvilvl.w    \tmp0,   \in6,    \in4
+    xvilvl.w    \tmp1,   \in7,    \in5
+    xvilvh.w    \tmp2,   \in6,    \in4
+    xvilvh.w    \tmp3,   \in7,    \in5
+    xvilvl.w    \out4,   \tmp1,   \tmp0
+    xvilvh.w    \out5,   \tmp1,   \tmp0
+    xvilvl.w    \out6,   \tmp3,   \tmp2
+    xvilvh.w    \out7,   \tmp3,   \tmp2
+
+    xmov        \tmp0,   \out0
+    xmov        \tmp1,   \out1
+    xmov        \tmp2,   \out2
+    xmov        \tmp3,   \out3
+    xvpermi.q   \out0,   \out4,   0x02
+    xvpermi.q   \out1,   \out5,   0x02
+    xvpermi.q   \out2,   \out6,   0x02
+    xvpermi.q   \out3,   \out7,   0x02
+    xvpermi.q   \out4,   \tmp0,   0x31
+    xvpermi.q   \out5,   \tmp1,   0x31
+    xvpermi.q   \out6,   \tmp2,   0x31
+    xvpermi.q   \out7,   \tmp3,   0x31
+.endm
+
+/*
+ * Description : Transpose 4x4 block with double-word elements in vectors
+ * Arguments   : Inputs  - in0, in1, in2, in3
+ *               Outputs - out0, out1, out2, out3
+ * Example     : LASX_TRANSPOSE4x4_D
+ *         in0 : 1,2,3,4
+ *         in1 : 1,2,3,4
+ *         in2 : 1,2,3,4
+ *         in3 : 1,2,3,4
+ *
+ *        out0 : 1,1,1,1
+ *        out1 : 2,2,2,2
+ *        out2 : 3,3,3,3
+ *        out3 : 4,4,4,4
+ */
+.macro LASX_TRANSPOSE4x4_D in0, in1, in2, in3, out0, out1, out2, out3, \
+                           tmp0, tmp1
+    xvilvl.d    \tmp0,   \in1,    \in0
+    xvilvh.d    \out1,   \in1,    \in0
+    xvilvh.d    \tmp1,   \in3,    \in2
+    xvilvl.d    \out2,   \in3,    \in2
+
+    xvor.v      \out0,   \tmp0,   \tmp0
+    xvor.v      \out3,   \tmp1,   \tmp1
+
+    xvpermi.q   \out0,   \out2,   0x02
+    xvpermi.q   \out2,   \tmp0,   0x31
+    xvpermi.q   \out3,   \out1,   0x31
+    xvpermi.q   \out1,   \tmp1,   0x02
+.endm
diff --git a/source/common/loongarch64/loopfilter.S b/source/common/loongarch64/loopfilter.S
new file mode 100644
index 000000000..42536616c
--- /dev/null
+++ b/source/common/loongarch64/loopfilter.S
@@ -0,0 +1,1324 @@
+/*
+*****************************************************************************
+* Copyright (C) 2013-2024 MulticoreWare, Inc
+*
+* Authors: Hao Chen <chenhao@loongson.cn>
+*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02111, USA.
+*
+* This program is also available under a commercial proprietary license.
+* For more information, contact us at license @ x265.com.
+*****************************************************************************
+*/
+
+#include "loongson_asm.S"
+
+const shuf
+.byte 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30
+endconst
+
+/*x265_saoCuOrgE0_lsx(pixel * rec, int8_t * offsetEo, int width,
+                      int8_t* signLeft, intptr_t stride)
+*/
+function x265_saoCuOrgE0_lsx
+    la.local     t2,    shuf
+    vldrepl.b    vr0,   a3,    0
+    vldi         vr10,  1
+    vld          vr11,  a1,    0
+    vld          vr12,  t2,    0
+    move         t0,    a0
+    move         t1,    a2
+.loop_e0:
+    vld          vr1,   t0,    0
+    vld          vr2,   t0,    1
+    vslt.bu      vr3,   vr2,   vr1
+    vslt.bu      vr4,   vr1,   vr2
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vshuf.b      vr7,   vr6,   vr0,   vr12
+    vadd.b       vr8,   vr7,   vr5
+    vreplvei.b   vr0,   vr6,   15
+    vaddi.bu     vr9,   vr8,   2
+    vshuf.b      vr7,   vr11,  vr11,  vr9
+    vaddwev.h.bu.b  vr8,   vr1,   vr7
+    vaddwod.h.bu.b  vr9,   vr1,   vr7
+    vssrani.bu.h vr9,   vr8,   0
+    vshuf4i.w    vr3,   vr9,   0x4E
+    vilvl.b      vr4,   vr3,   vr9
+    vst          vr4,   t0,    0
+    addi.d       t1,    t1,    -16
+    addi.d       t0,    t0,    16
+    bnez         t1,    .loop_e0
+    vldrepl.b    vr0,   a3,    1
+    add.d        t0,    a0,    a4
+.loopH_e0:
+    vld          vr1,   t0,    0
+    vld          vr2,   t0,    1
+    vslt.bu      vr3,   vr2,   vr1
+    vslt.bu      vr4,   vr1,   vr2
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vshuf.b      vr7,   vr6,   vr0,   vr12
+    vadd.b       vr8,   vr7,   vr5
+    vreplvei.b   vr0,   vr6,   15
+    vaddi.bu     vr9,   vr8,   2
+    vshuf.b      vr7,   vr11,  vr11,  vr9
+    vaddwev.h.bu.b  vr8,   vr1,   vr7
+    vaddwod.h.bu.b  vr9,   vr1,   vr7
+    vssrani.bu.h vr9,   vr8,   0
+    vshuf4i.w    vr3,   vr9,   0x4E
+    vilvl.b      vr4,   vr3,   vr9
+    vst          vr4,   t0,    0
+    addi.d       a2,    a2,    -16
+    addi.d       t0,    t0,    16
+    bnez         a2,    .loopH_e0
+endfunc
+
+/*x265_saoCuOrgE1_lsx(pixel* rec, int8_t* upBuff1,
+                      int8_t* offsetEo, intptr_t stride, int width)
+*/
+function x265_saoCuOrgE1_lsx
+    vldi         vr10,  1
+    vld          vr11,  a2,    0
+    srli.d       t0,    a4,    4
+.loop_e1:
+    vld          vr0,   a0,    0
+    vldx         vr1,   a0,    a3
+    vld          vr2,   a1,    0
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x4E
+    addi.d       t0,    t0,    -1
+    vilvl.b      vr7,   vr5,   vr4
+    vst          vr6,   a1,    0
+    vst          vr7,   a0,    0
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    16
+    bnez         t0,    .loop_e1
+endfunc
+
+/*x265_saoCuOrgE1_2Rows_lsx(pixel* rec, int8_t* upBuff1, int8_t* offsetEo,
+                            intptr_t stride, int width)
+*/
+function x265_saoCuOrgE1_2Rows_lsx
+    vldi         vr10,  1
+    vld          vr11,  a2,    0
+    srli.d       t0,    a4,    4
+    move         t1,    a0
+    move         t2,    a1
+    move         t3,    t0
+.loop_2rows_e1:
+    vld          vr0,   t1,    0
+    vldx         vr1,   t1,    a3
+    vld          vr2,   t2,    0
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x4E
+    addi.d       t0,    t0,    -1
+    vilvl.b      vr7,   vr5,   vr4
+    vst          vr6,   t2,    0
+    vst          vr7,   t1,    0
+    addi.d       t2,    t2,    16
+    addi.d       t1,    t1,    16
+    bnez         t0,    .loop_2rows_e1
+    add.d        a0,    a0,    a3
+.loopH_2rows_e1:
+    vld          vr0,   a0,    0
+    vldx         vr1,   a0,    a3
+    vld          vr2,   a1,    0
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x0E
+    addi.d       t3,    t3,    -1
+    vilvl.b      vr7,   vr5,   vr4
+    vst          vr6,   a1,    0
+    vst          vr7,   a0,    0
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    16
+    bnez         t3,    .loopH_2rows_e1
+endfunc
+
+/*x265_saoCuOrgE2_lsx(pixel * rec, int8_t * bufft, int8_t * buff1,
+                      int8_t * offsetEo, int width, intptr_t stride)
+*/
+function x265_saoCuOrgE2_lsx
+    vldi         vr10,  1
+    vld          vr11,  a3,    0
+    srli.d       t0,    a4,    4
+    andi         t1,    a4,    0xF
+    addi.d       a5,    a5,    1
+    addi.d       a1,    a1,    1
+    beqz         t0,    .loop_end_e2
+.loop_e2:
+    vld          vr0,   a0,    0
+    vldx         vr1,   a0,    a5
+    vld          vr2,   a2,    0
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x04E
+    addi.d       t0,    t0,    -1
+    vilvl.b      vr7,   vr5,   vr4
+    vst          vr6,   a1,    0
+    vst          vr7,   a0,    0
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    16
+    addi.d       a2,    a2,    16
+    blt          zero,  t0,    .loop_e2
+.loop_end_e2:
+    beqz         t1,    .end_e2
+    vld          vr0,   a0,    0
+    vldx         vr1,   a0,    a5
+    vld          vr2,   a2,    0
+    andi         t3,    t1,    0x8
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x4E
+    vilvl.b      vr7,   vr5,   vr4
+    beqz         t3,    .e2_res
+    fst.d        f6,    a1,    0
+    fst.d        f7,    a0,    0
+    addi.d       a1,    a1,    8
+    addi.d       a0,    a0,    8
+    addi.d       t1,    t1,    -8
+    vshuf4i.w    vr6,   vr6,   0x4E
+    vshuf4i.w    vr7,   vr7,   0x4E
+    beqz         t1,    .end_e2
+.e2_res:
+    vilvl.d      vr7,   vr7,   vr6
+.loop_e2_res:
+    vstelm.b     vr7,   a1,    0,    0
+    vstelm.b     vr7,   a0,    0,    8
+    addi.d       a1,    a1,    1
+    addi.d       a0,    a0,    1
+    addi.d       t1,    t1,    -1
+    vbsrl.v      vr7,   vr7,   1
+    blt          zero,  t1,    .loop_e2_res
+.end_e2:
+endfunc
+
+/*x265_saoCuOrgE3_lsx(pixel *rec, int8_t *upBuff1, int8_t *offsetEo,
+                      intptr_t stride, int startX, int endX)
+*/
+function x265_saoCuOrgE3_lsx
+    add.d        a1,    a1,    a4
+    addi.d       a4,    a4,    1
+    vldi         vr10,  1
+    vld          vr11,  a2,    0
+    sub.d        t0,    a5,    a4
+    add.d        a0,    a0,    a4
+    srli.d       t1,    t0,    4
+    andi         t2,    t0,    0xF
+    beqz         t1,    .loop_end_e3
+.loop_e3:
+    vld          vr0,   a0,    0
+    vldx         vr1,   a0,    a3
+    vld          vr2,   a1,    1
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x04E
+    addi.d       t1,    t1,    -1
+    vilvl.b      vr7,   vr5,   vr4
+    vst          vr6,   a1,    0
+    vst          vr7,   a0,    0
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    16
+    blt          zero,  t1,    .loop_e3
+.loop_end_e3:
+    beqz         t2,    .end_e3
+    vld          vr0,   a0,    0
+    vldx         vr1,   a0,    a3
+    vld          vr2,   a1,    1
+    andi         t3,    t2,    0x8
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vshuf.b      vr9,   vr11,  vr11,   vr8
+    vaddwev.h.bu.b  vr3,   vr0,   vr9
+    vaddwod.h.bu.b  vr4,   vr0,   vr9
+    vssrani.bu.h vr4,   vr3,   0
+    vshuf4i.w    vr5,   vr4,   0x4E
+    vilvl.b      vr7,   vr5,   vr4
+    beqz         t3,    .e3_res
+    fst.d        f6,    a1,    0
+    fst.d        f7,    a0,    0
+    addi.d       a1,    a1,    8
+    addi.d       a0,    a0,    8
+    addi.d       t2,    t2,    -8
+    vshuf4i.w    vr6,   vr6,   0x4E
+    vshuf4i.w    vr7,   vr7,   0x4E
+    beqz         t2,    .end_e3
+.e3_res:
+    vilvl.d      vr7,   vr7,   vr6
+.loop_e3_res:
+    vstelm.b     vr7,   a1,    0,    0
+    vstelm.b     vr7,   a0,    0,    8
+    addi.d       a1,    a1,    1
+    addi.d       a0,    a0,    1
+    addi.d       t2,    t2,    -1
+    vbsrl.v      vr7,   vr7,   1
+    blt          zero,  t2,    .loop_e3_res
+.end_e3:
+endfunc
+
+/*x265_saoCuOrgB0_lsx(pixel* rec, const int8_t* offset, int ctuWidth,
+                      int ctuHeight, intptr_t stride)
+*/
+function x265_saoCuOrgB0_lsx
+    srli.d       t0,    a3,    1
+    srli.d       t1,    a2,    4
+    slli.d       t2,    a4,    1
+    andi         t3,    a3,    1
+    vld          vr0,   a1,    0
+    vld          vr1,   a1,    16
+    beqz         t0,    .loopH_b0_end
+.loopH_b0:
+    move         t4,    a0
+    move         t5,    t1
+.loopW_b0:
+    vld          vr2,   t4,    0
+    vldx         vr3,   t4,    a4
+    vsrli.b      vr4,   vr2,   3
+    vsrli.b      vr5,   vr3,   3
+    vshuf.b      vr6,   vr1,   vr0,  vr4
+    vshuf.b      vr7,   vr1,   vr0,  vr5
+    vaddwev.h.bu.b  vr8,  vr2,  vr6
+    vaddwev.h.bu.b  vr10, vr3,  vr7
+    vaddwod.h.bu.b  vr9,  vr2,  vr6
+    vaddwod.h.bu.b  vr11, vr3,  vr7
+    vssrani.bu.h vr10,  vr8,   0
+    vssrani.bu.h vr11,  vr9,   0
+    vilvl.b      vr12,  vr11,  vr10
+    vilvh.b      vr13,  vr11,  vr10
+    vst          vr12,  t4,    0
+    vstx         vr13,  t4,    a4
+    addi.d       t5,    t5,    -1
+    addi.d       t4,    t4,    16
+    blt          zero,  t5,    .loopW_b0
+    addi.d       t0,    t0,    -1
+    add.d        a0,    a0,    t2
+    blt          zero,  t0,    .loopH_b0
+.loopH_b0_end:
+    beqz         t3,    .end_b0
+.loopW1_b0:
+    vld          vr2,   a0,    0
+    vsrli.b      vr4,   vr2,   3
+    vshuf.b      vr6,   vr1,   vr0,  vr4
+    vaddwev.h.bu.b  vr8,  vr2,  vr6
+    vaddwod.h.bu.b  vr10, vr2,  vr6
+    vssrani.bu.h vr10,  vr8,   0
+    vshuf4i.w    vr11,  vr10,  0x4E
+    vilvl.b      vr12,  vr11,  vr10
+    vst          vr12,  a0,    0
+    addi.d       t1,    t1,    -1
+    addi.d       a0,    a0,    16
+    blt          zero,  t1,    .loopW1_b0
+.end_b0:
+endfunc
+
+/* x265_calSign_lsx(int8_t *dst, const pixel *src1, const pixel *src2, const int endX) */
+function x265_calSign_lsx
+    srli.d       t0,    a3,    4
+    andi         t1,    a3,    15
+    vldi         vr10,  1
+    beqz         t0,    .loop_sign_lsx_end
+.loop_sign_lsx:
+    vld          vr0,   a1,    0
+    vld          vr1,   a2,    0
+    addi.d       t0,    t0,    -1
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vst          vr5,   a0,    0
+    addi.d       a1,    a1,    16
+    addi.d       a2,    a2,    16
+    addi.d       a0,    a0,    16
+    blt          zero,  t0,    .loop_sign_lsx
+.loop_sign_lsx_end:
+    beqz         t1,    .end_sign_lsx
+    vld          vr0,   a1,    0
+    vld          vr1,   a2,    0
+    andi         t2,    t1,    8
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    beqz         t2,    .loop_sign_res_lsx
+    fst.d        f5,    a0,    0
+    addi.d       t1,    t1,    -8
+    addi.d       a0,    a0,    8
+    vshuf4i.w    vr5,   vr5,   0x4E
+    beqz         t1,    .end_sign_lsx
+.loop_sign_res_lsx:
+    vstelm.b     vr5,   a0,    0,    0
+    addi.d       t1,    t1,    -1
+    addi.d       a0,    a0,    1
+    vbsrl.v      vr5,   vr5,   1
+    blt          zero,  t1,    .loop_sign_res_lsx
+.end_sign_lsx:
+endfunc
+
+/* x265_calSign_lasx(int8_t *dst, const pixel *src1, const pixel *src2, const int endX) */
+function x265_calSign_lasx
+    srli.d       t0,    a3,    5
+    andi         t1,    a3,    31
+    xvldi        xr10,  1
+    beqz         t0,    .loop_sign_lasx_end
+.loop_sign_lasx:
+    xvld         xr0,   a1,    0
+    xvld         xr1,   a2,    0
+    addi.d       t0,    t0,    -1
+    xvslt.bu     xr3,   xr1,   xr0
+    xvslt.bu     xr4,   xr0,   xr1
+    xvand.v      xr3,   xr3,   xr10
+    xvor.v       xr5,   xr3,   xr4
+    xvst         xr5,   a0,    0
+    addi.d       a1,    a1,    32
+    addi.d       a2,    a2,    32
+    addi.d       a0,    a0,    32
+    blt          zero,  t0,    .loop_sign_lasx
+.loop_sign_lasx_end:
+    beqz         t1,    .end_sign_lasx
+    xvld         xr0,   a1,    0
+    xvld         xr1,   a2,    0
+    andi         t2,    t1,    16
+    xvslt.bu     xr3,   xr1,   xr0
+    xvslt.bu     xr4,   xr0,   xr1
+    xvand.v      xr3,   xr3,   xr10
+    xvor.v       xr5,   xr3,   xr4
+    beqz         t2,    .loop_sign_res_lasx
+    vst          vr5,   a0,    0
+    addi.d       t1,    t1,    -16
+    addi.d       a0,    a0,    16
+    xvpermi.d    xr5,   xr5,   0x4E
+    beqz         t1,    .end_sign_lasx
+.loop_sign_res_lasx:
+    vstelm.b     vr5,   a0,    0,    0
+    addi.d       t1,    t1,    -1
+    addi.d       a0,    a0,    1
+    vbsrl.v      vr5,   vr5,   1
+    blt          zero,  t1,    .loop_sign_res_lasx
+.end_sign_lasx:
+endfunc
+
+/* x265_pelFilterLumaStrong_V_lsx(pixel* src, intptr_t srcStep, intptr_t offset,
+                                  int32_t tcP, int32_t tcQ)
+*/
+function x265_pelFilterLumaStrong_V_lsx
+    addi.d       a0,    a0,    -3
+    vreplgr2vr.h vr0,   a3   //tcP
+    vreplgr2vr.h vr1,   a4   //tcQ
+    add.d        t0,    a0,    a1
+    add.d        t1,    t0,    a1
+    add.d        t2,    t1,    a1
+    vilvl.d      vr0,   vr1,   vr0  //[tcP, tcQ]
+    vldi         vr10,  0
+    vneg.h       vr1,   vr0         //[-tcP, -tcQ]
+
+    fld.d        f2,    a0,    -1   //src[m0,m1, m2, m3, m4, m5,m6, m7]
+    fld.d        f3,    t0,    -1
+    fld.d        f4,    t1,    -1
+    fld.d        f5,    t2,    -1
+
+    vilvl.b      vr6,   vr3,   vr2 //[m0, m0, ... m3, m3, m4, m4 ... m7, m7]
+    vilvl.b      vr7,   vr5,   vr4 //[m0, m0]
+
+    vpackev.h    vr2,   vr7,   vr6 //[m0, m2, m4, m6]
+    vpackod.h    vr3,   vr7,   vr6 //[m1, m3, m5, m7]
+
+    vilvl.b      vr4,   vr10,  vr2  //[m0, m2]
+    vilvh.b      vr5,   vr10,  vr2  //[m4, m6]
+    vilvl.b      vr6,   vr10,  vr3  //[m1, m3]
+    vilvh.b      vr7,   vr10,  vr3  //[m5, m7]
+
+    vadd.h       vr8,   vr4,   vr6  //[m0+m1, m2+m3]
+    vadd.h       vr9,   vr5,   vr7  //[m4+m5, m6+m7]
+
+    vshuf4i.w    vr11,  vr8,   0x4E //[m2+m3, m0+m1]
+    vslli.h      vr14,  vr8,   1    //2*[m0+m1, m2+m3]
+    vslli.h      vr15,  vr9,   1    //2*[m4+m5, m6+m7]
+    vadd.h       vr2,   vr5,   vr6  //[m1+m4, m6+m3]
+    vadd.h       vr12,  vr9,   vr11 //[m2+m3+m4+m5, m0+m1+m6+m7]
+    vpermi.w     vr15,  vr14,  0xE4 //2*[m0+m1, m6+m7]
+
+    vpermi.w     vr9,   vr8,   0x4E //[m2+m3, m4+m5]
+    vpermi.w     vr7,   vr4,   0x4E //[m2, m5]
+    vshuf4i.w    vr5,   vr5,   0x4E //[m6, m4]
+    vshuf4i.w    vr12,  vr12,  0x44 //[m2+m3+m4+m5, m2+m3+m4+m5]
+    vadd.h       vr3,   vr2,   vr9  //[m1+m2+m3+m4, m3+m4+m5+m6]
+    vilvl.d      vr11,  vr5,   vr6  //[m1, m6]
+    vilvh.d      vr13,  vr5,   vr6  //[m3, m4]
+    vsrlri.h     vr4,   vr3,   2    //([m1+m2+m3+m4, m3+m4+m5+m6]+2)>>2
+    vadd.h       vr5,   vr15,  vr3  //[2m0+2m1+m1+m2+m3+m4, m3+m4+m5+m6+2m6+2m7]
+    vadd.h       vr6,   vr12,  vr3  //[m1+2m2+2m3+2m4+m5, m2+2m3+2m4+2m5+m6]
+    vsrlri.h     vr5,   vr5,   3
+    vsrlri.h     vr6,   vr6,   3
+    vsub.h       vr4,   vr4,   vr7  //([m1+m2+m3+m4, m3+m4+m5+m6]+2)>>2 - [m2, m5]
+    vsub.h       vr5,   vr5,   vr11 //([2m0+2m1+m1+m2+m3+m4, m3+m4+m5+m6+2m6+2m7]+4)>>3 - [m1, m6]
+    vsub.h       vr6,   vr6,   vr13 //([m1+2m2+2m3+2m4+m5, m2+2m3+2m4+2m5+m6]+4)>>3 - [m3, m4]
+    vclip.h      vr3,   vr4,   vr1,   vr0  //[-2, 1]
+    vclip.h      vr2,   vr5,   vr1,   vr0  //[-3, 2]
+    vclip.h      vr4,   vr6,   vr1,   vr0  //[-1, 0]
+
+    vadd.h       vr2,   vr2,   vr11   //[-3, 2]
+    vadd.h       vr3,   vr3,   vr7    //[-2, 1]
+    vadd.h       vr4,   vr4,   vr13   //[-1, 0]
+    vpackev.b    vr5,   vr3,   vr2    //[-3,-2, 2, 1]
+    vshuf4i.w    vr14,  vr4,   0x4E   //[0, -1]
+    vpackev.b    vr6,   vr2,   vr3    //[-2,-3, 1, 2]
+    vpackev.b    vr7,   vr14,  vr4    //[-1, 0, 0, -1]
+    vilvl.h      vr8,   vr7,   vr5    //[-3,-2,-1,0]
+    vstelm.w     vr8,   a0,    0,     0
+    vstelm.w     vr8,   t0,    0,     1
+    vstelm.w     vr8,   t1,    0,     2
+    vstelm.w     vr8,   t2,    0,     3
+    vstelm.h     vr6,   a0,    4,     4
+    vstelm.h     vr6,   t0,    4,     5
+    vstelm.h     vr6,   t1,    4,     6
+    vstelm.h     vr6,   t2,    4,     7
+endfunc
+
+function x265_pelFilterLumaStrong_H_lsx
+    slli.d       t0,    a2,    1    //offset * 2
+    slli.d       t2,    a2,    2    //offset * 4
+    add.d        t1,    a2,    t0   //offset * 3
+    vreplgr2vr.h vr0,   a3   //tcP
+    vreplgr2vr.h vr1,   a4   //tcQ
+    sub.d        t3,    a0,    t2
+    vilvl.d      vr0,   vr1,   vr0  //[tcP, tcQ]
+    vldi         vr10,  0
+    vneg.h       vr1,   vr0         //[-tcP, -tcQ]
+
+    fld.s        f2,    t3,    0    //[m0]
+    fldx.s       f3,    t3,    a2   //[m1]
+    fldx.s       f4,    t3,    t0   //[m2]
+    fldx.s       f5,    t3,    t1   //[m3]
+    fld.s        f6,    a0,    0    //[m4]
+    fldx.s       f7,    a0,    a2   //[m5]
+    fldx.s       f8,    a0,    t0   //[m6]
+    fldx.s       f9,    a0,    t1   //[m7]
+
+    vilvl.w      vr11,  vr4,   vr2  //[m0, m2]
+    vilvl.w      vr12,  vr8,   vr6  //[m4, m6]
+    vilvl.w      vr13,  vr5,   vr3  //[m1, m3]
+    vilvl.w      vr14,  vr9,   vr7  //[m5, m7]
+
+    vilvl.b      vr4,   vr10,  vr11  //[m0, m2]
+    vilvl.b      vr5,   vr10,  vr12  //[m4, m6]
+    vilvl.b      vr6,   vr10,  vr13  //[m1, m3]
+    vilvl.b      vr7,   vr10,  vr14  //[m5, m7]
+
+    vadd.h       vr8,   vr4,   vr6  //[m0+m1, m2+m3]
+    vadd.h       vr9,   vr5,   vr7  //[m4+m5, m6+m7]
+
+    vshuf4i.w    vr11,  vr8,   0x4E //[m2+m3, m0+m1]
+    vslli.h      vr14,  vr8,   1    //2*[m0+m1, m2+m3]
+    vslli.h      vr15,  vr9,   1    //2*[m4+m5, m6+m7]
+    vadd.h       vr2,   vr5,   vr6  //[m1+m4, m6+m3]
+    vadd.h       vr12,  vr9,   vr11 //[m2+m3+m4+m5, m0+m1+m6+m7]
+    vpermi.w     vr15,  vr14,  0xE4 //2*[m0+m1, m6+m7]
+
+    vpermi.w     vr9,   vr8,   0x4E //[m2+m3, m4+m5]
+    vpermi.w     vr7,   vr4,   0x4E //[m2, m5]
+    vshuf4i.w    vr5,   vr5,   0x4E //[m6, m4]
+    vshuf4i.w    vr12,  vr12,  0x44 //[m2+m3+m4+m5, m2+m3+m4+m5]
+    vadd.h       vr3,   vr2,   vr9  //[m1+m2+m3+m4, m3+m4+m5+m6]
+    vilvl.d      vr11,  vr5,   vr6  //[m1, m6]
+    vilvh.d      vr13,  vr5,   vr6  //[m3, m4]
+    vsrlri.h     vr4,   vr3,   2    //([m1+m2+m3+m4, m3+m4+m5+m6]+2)>>2
+    vadd.h       vr5,   vr15,  vr3  //[2m0+2m1+m1+m2+m3+m4, m3+m4+m5+m6+2m6+2m7]
+    vadd.h       vr6,   vr12,  vr3  //[m1+2m2+2m3+2m4+m5, m2+2m3+2m4+2m5+m6]
+    vsrlri.h     vr5,   vr5,   3
+    vsrlri.h     vr6,   vr6,   3
+    vsub.h       vr4,   vr4,   vr7  //([m1+m2+m3+m4, m3+m4+m5+m6]+2)>>2 - [m2, m5]
+    vsub.h       vr5,   vr5,   vr11 //([2m0+2m1+m1+m2+m3+m4, m3+m4+m5+m6+2m6+2m7]+4)>>3 - [m1, m6]
+    vsub.h       vr6,   vr6,   vr13 //([m1+2m2+2m3+2m4+m5, m2+2m3+2m4+2m5+m6]+4)>>3 - [m3, m4]
+    vclip.h      vr3,   vr4,   vr1,   vr0  //[-2, 1]
+    vclip.h      vr2,   vr5,   vr1,   vr0  //[-3, 2]
+    vclip.h      vr4,   vr6,   vr1,   vr0  //[-1, 0]
+
+    vadd.h       vr2,   vr2,   vr11   //[-3, 2]
+    vadd.h       vr3,   vr3,   vr7    //[-2, 1]
+    vadd.h       vr4,   vr4,   vr13   //[-1, 0]
+
+    sub.d        t3,    a0,    t1   //-offset*3
+    sub.d        t4,    a0,    t0   //-offset*2
+    sub.d        t5,    a0,    a2   //-offset
+    add.d        t6,    a0,    a2   //offset
+    add.d        t7,    a0,    t0   //offset*2
+    vpickev.b    vr5,   vr3,   vr2   //[-3, 2, -2, 1]
+    vpickev.b    vr6,   vr10,  vr4   //[-1, 0]
+
+    vstelm.w     vr5,   t3,    0,    0
+    vstelm.w     vr5,   t4,    0,    2
+    vstelm.w     vr5,   t6,    0,    3
+    vstelm.w     vr5,   t7,    0,    1
+    vstelm.w     vr6,   t5,    0,    0
+    vstelm.w     vr6,   a0,    0,    1
+endfunc
+
+/*x265_pelFilterChroma_V_lsx(pixel* src, intptr_t srcStep, intptr_t offset,
+                             int32_t tc, int32_t maskP, int32_t maskQ)
+*/
+function x265_pelFilterChroma_V_lsx
+    addi.d       a0,    a0,    -1
+    vreplgr2vr.h vr0,   a3   //tcP
+    vldi         vr10,  0
+    add.d        t0,    a0,    a1
+    add.d        t1,    t0,    a1
+    add.d        t2,    t1,    a1
+    vneg.h       vr1,   vr0         //[-tcP, -tcQ]
+    vreplgr2vr.h vr2,   a4   //maskP
+    vreplgr2vr.h vr3,   a5   //maskQ
+
+    fld.s        f4,    a0,    -1   //src[m2, m3, m4, m5]
+    fld.s        f5,    t0,    -1
+    fld.s        f6,    t1,    -1
+    fld.s        f7,    t2,    -1
+
+    vilvl.b      vr8,   vr5,   vr4   //[m2, m2, m3, m3, m4, m4, m5, m5]
+    vilvl.b      vr9,   vr7,   vr6
+    vilvl.h      vr11,  vr9,   vr8   //[m2, m3, m4, m5]
+
+    vilvl.b      vr4,   vr10,  vr11  //[m2, m3]
+    vilvh.b      vr6,   vr10,  vr11  //[m4, m5]
+    vbsrl.v      vr5,   vr4,   8     //[m3]
+    vbsrl.v      vr7,   vr6,   8     //[m5]
+
+    vsub.h       vr8,   vr6,   vr5   //[m4-m3]
+    vsub.h       vr9,   vr4,   vr7   //[m2-m5]
+    vslli.h      vr8,   vr8,   2     //[m4-m3] * 4
+    vadd.h       vr11,  vr8,   vr9
+    vsrari.h     vr12,  vr11,  3
+    vclip.h      vr13,  vr12,  vr1,  vr0  //[delta]
+    vand.v       vr14,  vr13,  vr2   //-offset
+    vand.v       vr15,  vr13,  vr3   //0
+    vadd.h       vr8,   vr5,   vr14
+    vsub.h       vr9,   vr6,   vr15
+
+    vilvl.h      vr11,  vr9,   vr8
+    vmaxi.h      vr13,  vr11,  0
+    vsat.hu      vr14,  vr13,  7
+    vpickev.b    vr12,  vr10,  vr14
+    vstelm.h     vr12,  a0,    0,     0
+    vstelm.h     vr12,  t0,    0,     1
+    vstelm.h     vr12,  t1,    0,     2
+    vstelm.h     vr12,  t2,    0,     3
+endfunc
+
+function x265_pelFilterChroma_H_lsx
+    slli.d       t0,    a2,    1    //offset * 2
+    vreplgr2vr.h vr0,   a3   //tcP
+    vldi         vr10,  0
+    sub.d        t1,    a0,    t0
+    vneg.h       vr1,   vr0         //[-tcP, -tcQ]
+    vreplgr2vr.h vr2,   a4   //maskP
+    vreplgr2vr.h vr3,   a5   //maskQ
+
+    fld.s        f4,    t1,    0   //src[m2]
+    fldx.s       f5,    t1,    a2  //src[m3]
+    fld.s        f6,    a0,    0   //src[m4]
+    fldx.s       f7,    a0,    a2  //src[m5]
+
+    vilvl.b      vr4,   vr10,  vr4
+    vilvl.b      vr5,   vr10,  vr5
+    vilvl.b      vr6,   vr10,  vr6
+    vilvl.b      vr7,   vr10,  vr7
+
+    vsub.h       vr8,   vr6,   vr5   //[m4-m3]
+    vsub.h       vr9,   vr4,   vr7   //[m2-m5]
+    vslli.h      vr8,   vr8,   2     //[m4-m3] * 4
+    vadd.h       vr11,  vr8,   vr9
+    vsrari.h     vr12,  vr11,  3
+    vclip.h      vr13,  vr12,  vr1,  vr0  //[delta]
+    vand.v       vr14,  vr13,  vr2   //-offset
+    vand.v       vr15,  vr13,  vr3   //0
+    vadd.h       vr8,   vr5,   vr14
+    vsub.h       vr9,   vr6,   vr15
+
+    vssrani.bu.h vr9,   vr8,   0
+    fstx.s       f9,    t1,    a2
+    vstelm.w     vr9,   a0,    0,    2
+endfunc
+
+.macro STATS_COUNT in0, in1, i0, i1
+    vpickve2gr.b t4,    \in0,  \i0
+    vpickve2gr.h t8,    \in1,  \i1
+    ldx.w        t6,    sp,    t4
+    ldx.w        t7,    t5,    t4
+    add.w        t6,    t6,    t8
+    addi.w       t7,    t7,    1
+    stx.w        t6,    sp,    t4
+    stx.w        t7,    t5,    t4
+.endm
+
+/*x265_saoCuStatsE0_lsx(const int16_t *diff, const pixel *rec, intptr_t stride, int endX,
+                        int endY, int32_t *stats, int32_t *count)
+*/
+function x265_saoCuStatsE0_lsx
+    la.local     t2,    shuf
+    addi.d       sp,    sp,    -48
+    vldi         vr10,  1
+    vldi         vr11,  0
+    vld          vr12,  t2,    0
+    addi.d       t5,    sp,    20
+    vst          vr11,  sp,    0
+    vst          vr11,  sp,    16
+    vst          vr11,  sp,    32
+.loopS_Y_e0:
+    ld.bu        t3,    a1,    0
+    ld.bu        t4,    a1,    -1
+    addi.d       a7,    zero,  -1
+    sltu         t7,    t3,    t4   //rec[0] < rec[-1] ? 1:0
+    sltu         t6,    t4,    t3   //rec[-1] < rec[0] ? 1:0
+    maskeqz      t8,    a7,    t7
+    or           t6,    t6,    t8
+    vreplgr2vr.b vr0,   t6
+    srli.d       t0,    a3,    4
+    move         t3,    a1
+    move         t2,    a0
+    beqz         t0,    .ResS_X_e0
+.loopS_X_e0:
+    vld          vr1,   t3,    0
+    vld          vr2,   t3,    1
+    vld          vr13,  t2,    0
+    vld          vr14,  t2,    16
+    vslt.bu      vr3,   vr2,   vr1
+    vslt.bu      vr4,   vr1,   vr2
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4   //signRight
+    vneg.b       vr6,   vr5          //signLeft
+    vshuf.b      vr7,   vr6,   vr0,   vr12
+    vadd.b       vr8,   vr7,   vr5
+    vreplvei.b   vr0,   vr6,   15
+    vaddi.bu     vr9,   vr8,   2    //edgeType
+    vslli.b      vr9,   vr9,   2
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    STATS_COUNT vr9, vr14, 8,  0
+    STATS_COUNT vr9, vr14, 9,  1
+    STATS_COUNT vr9, vr14, 10, 2
+    STATS_COUNT vr9, vr14, 11, 3
+    STATS_COUNT vr9, vr14, 12, 4
+    STATS_COUNT vr9, vr14, 13, 5
+    STATS_COUNT vr9, vr14, 14, 6
+    STATS_COUNT vr9, vr14, 15, 7
+    addi.d       t0,    t0,    -1
+    addi.d       t3,    t3,    16
+    addi.d       t2,    t2,    32
+    bnez         t0,    .loopS_X_e0
+.ResS_X_e0:
+    andi         t1,    a3,    15
+    beqz         t1,    .endS_X_e0
+    addi.d       t4,    zero,  8
+    vld          vr1,   t3,    0
+    vld          vr2,   t3,    1
+    vld          vr13,  t2,    0
+    vld          vr14,  t2,    16
+    vslt.bu      vr3,   vr2,   vr1
+    vslt.bu      vr4,   vr1,   vr2
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4   //signRight
+    vneg.b       vr6,   vr5          //signLeft
+    vshuf.b      vr7,   vr6,   vr0,   vr12
+    vadd.b       vr8,   vr7,   vr5
+    vreplvei.b   vr0,   vr6,   15
+    vaddi.bu     vr9,   vr8,   2    //edgeType
+    vslli.b      vr9,   vr9,   2
+    blt          t1,    t4,    .resS_X_e0
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    addi.d       t1,    t1,    -8
+    beqz         t1,    .endS_X_e0
+    vmov         vr13,  vr14
+    vbsrl.v      vr9,   vr9,   8
+.resS_X_e0:
+    vpickve2gr.b t4,    vr9,   0
+    vpickve2gr.h t8,    vr13,  0
+    ldx.w        t6,    sp,    t4
+    ldx.w        t7,    t5,    t4
+    add.w        t6,    t6,    t8
+    addi.w       t7,    t7,    1
+    stx.w        t6,    sp,    t4
+    stx.w        t7,    t5,    t4
+    vbsrl.v      vr9,   vr9,   1
+    vbsrl.v      vr13,  vr13,  2
+    addi.d       t1,    t1,    -1
+    bnez         t1,    .resS_X_e0
+.endS_X_e0:
+    addi.d       a4,    a4,    -1
+    add.d        a1,    a1,    a2
+    addi.d       a0,    a0,    128
+    bnez         a4,    .loopS_Y_e0
+    vld          vr1,   sp,    0
+    vld          vr2,   t5,    0
+    vld          vr3,   a5,    0
+    vld          vr4,   a6,    0
+    vshuf4i.w    vr5,   vr1,   0xD2
+    vshuf4i.w    vr6,   vr2,   0xD2
+    vadd.w       vr5,   vr5,   vr3
+    vadd.w       vr6,   vr6,   vr4
+    ld.w         t2,    a5,    16
+    ld.w         t3,    a6,    16
+    ld.w         t4,    sp,    16
+    ld.w         t6,    t5,    16
+    add.w        t2,    t2,    t4
+    add.w        t3,    t3,    t6
+    vst          vr5,   a5,    0
+    vst          vr6,   a6,    0
+    st.w         t2,    a5,    16
+    st.w         t3,    a6,    16
+    addi.d       sp,    sp,    48
+endfunc
+
+/*x265_saoCuStatsE1_lsx(const int16_t *diff, const pixel *rec, intptr_t stride, int8_t *upBuff1,
+                        int endx, int endY, int32_t *stats, int32_t *count)
+*/
+function x265_saoCuStatsE1_lsx
+    addi.d       sp,    sp,    -64
+    srli.d       t2,    a4,    4
+    andi         t3,    a4,    15
+    vldi         vr10,  1
+    vldi         vr11,  0
+    vst          vr11,  sp,    0
+    vst          vr11,  sp,    16
+    vst          vr11,  sp,    32
+    st.d         a0,    sp,    40
+    st.d         a1,    sp,    48
+    st.d         a3,    sp,    56
+    addi.d       t5,    sp,    20
+.loopS_Y_e1:
+    ld.d         a0,    sp,    40
+    ld.d         a1,    sp,    48
+    ld.d         a3,    sp,    56
+    move         t0,    t2
+    move         t1,    t3
+    beqz         t0,    .ResS_X_e1
+.loopS_X_e1:
+    vld          vr0,   a1,    0
+    vldx         vr1,   a1,    a2
+    vld          vr2,   a3,    0
+    vld          vr13,  a0,    0
+    vld          vr14,  a0,    16
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vst          vr6,   a3,    0
+    vslli.b      vr9,   vr8,   2
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    STATS_COUNT vr9, vr14, 8,  0
+    STATS_COUNT vr9, vr14, 9,  1
+    STATS_COUNT vr9, vr14, 10, 2
+    STATS_COUNT vr9, vr14, 11, 3
+    STATS_COUNT vr9, vr14, 12, 4
+    STATS_COUNT vr9, vr14, 13, 5
+    STATS_COUNT vr9, vr14, 14, 6
+    STATS_COUNT vr9, vr14, 15, 7
+    addi.d       t0,    t0,    -1
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    32
+    addi.d       a3,    a3,    16
+    bnez         t0,    .loopS_X_e1
+.ResS_X_e1:
+    beqz         t1,    .endS_X_e1
+    addi.d       t4,    zero,  8
+    vld          vr0,   a1,    0
+    vldx         vr1,   a1,    a2
+    vld          vr2,   a3,    0
+    vld          vr13,  a0,    0
+    vld          vr14,  a0,    16
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vslli.b      vr9,   vr8,   2
+    blt          t1,    t4,    .resS_X_e1
+    fst.d        f6,    a3,    0
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    addi.d       t1,    t1,    -8
+    addi.d       a3,    a3,    8
+    beqz         t1,    .endS_X_e1
+    vmov         vr13,  vr14
+    vbsrl.v      vr9,   vr9,   8
+    vbsrl.v      vr6,   vr6,   8
+.resS_X_e1:
+    vpickve2gr.b t4,    vr9,   0
+    vpickve2gr.h t8,    vr13,  0
+    ldx.w        t6,    sp,    t4
+    ldx.w        t7,    t5,    t4
+    add.w        t6,    t6,    t8
+    addi.w       t7,    t7,    1
+    stx.w        t6,    sp,    t4
+    stx.w        t7,    t5,    t4
+    vstelm.b     vr6,   a3,    0,   0
+    vbsrl.v      vr9,   vr9,   1
+    vbsrl.v      vr13,  vr13,  2
+    vbsrl.v      vr6,   vr6,   1
+    addi.d       t1,    t1,    -1
+    addi.d       a3,    a3,    1
+    bnez         t1,    .resS_X_e1
+.endS_X_e1:
+    ld.d         a0,    sp,    40
+    ld.d         a1,    sp,    48
+    addi.d       a5,    a5,    -1
+    addi.d       a0,    a0,    128
+    add.d        a1,    a1,    a2
+    st.d         a0,    sp,    40
+    st.d         a1,    sp,    48
+    bnez         a5,    .loopS_Y_e1
+    vld          vr1,   sp,    0
+    vld          vr2,   t5,    0
+    vld          vr3,   a6,    0
+    vld          vr4,   a7,    0
+    vshuf4i.w    vr5,   vr1,   0xD2
+    vshuf4i.w    vr6,   vr2,   0xD2
+    vadd.w       vr5,   vr5,   vr3
+    vadd.w       vr6,   vr6,   vr4
+    ld.w         t2,    a6,    16
+    ld.w         t3,    a7,    16
+    ld.w         t4,    sp,    16
+    ld.w         t6,    t5,    16
+    add.w        t2,    t2,    t4
+    add.w        t3,    t3,    t6
+    vst          vr5,   a6,    0
+    vst          vr6,   a7,    0
+    st.w         t2,    a6,    16
+    st.w         t3,    a7,    16
+    addi.d       sp,    sp,    64
+endfunc
+
+/*x265_saoCuStatsE2_lsx(const int16_t *diff, const pixel *rec, intptr_t stride, int8_t *upBuff1,
+                        int8_t *upBufft, int endX, int endY, int32_t *stats, int32_t *count)
+*/
+function x265_saoCuStatsE2_lsx
+    addi.d       sp,    sp,    -72
+    srli.d       t2,    a5,    4
+    andi         t3,    a5,    15
+    addi.d       a5,    a2,    1
+    vldi         vr10,  1
+    vldi         vr11,  0
+    vst          vr11,  sp,    0
+    vst          vr11,  sp,    16
+    vst          vr11,  sp,    32
+    st.d         a0,    sp,    40
+    st.d         a1,    sp,    48
+    st.d         a3,    sp,    56  //upBuff1
+    st.d         a4,    sp,    64  //upBufft
+    addi.d       t5,    sp,    20
+.loopS_Y_e2:
+    ld.d         a0,    sp,    40
+    ld.d         a1,    sp,    48
+    ld.d         a3,    sp,    56
+    ld.d         a4,    sp,    64
+    ldx.bu       t0,    a1,    a2
+    ld.bu        t1,    a1,    -1
+    addi.d       t4,    zero,  -1
+    sltu         t7,    t0,    t1
+    sltu         t6,    t1,    t0
+    maskeqz      t8,    t4,    t7
+    or           t6,    t6,    t8
+    st.b         t6,    a4,    0
+    move         t0,    t2
+    move         t1,    t3
+    beqz         t0,    .ResS_X_e2
+.loopS_X_e2:
+    vld          vr0,   a1,    0
+    vldx         vr1,   a1,    a5
+    vld          vr13,  a0,    0
+    vld          vr14,  a0,    16
+    vld          vr2,   a3,    0
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vst          vr6,   a4,    1
+    vslli.b      vr9,   vr8,   2
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    STATS_COUNT vr9, vr14, 8,  0
+    STATS_COUNT vr9, vr14, 9,  1
+    STATS_COUNT vr9, vr14, 10, 2
+    STATS_COUNT vr9, vr14, 11, 3
+    STATS_COUNT vr9, vr14, 12, 4
+    STATS_COUNT vr9, vr14, 13, 5
+    STATS_COUNT vr9, vr14, 14, 6
+    STATS_COUNT vr9, vr14, 15, 7
+    addi.d       t0,    t0,    -1
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    32
+    addi.d       a3,    a3,    16
+    addi.d       a4,    a4,    16
+    bnez         t0,    .loopS_X_e2
+.ResS_X_e2:
+    beqz         t1,    .endS_X_e2
+    addi.d       t4,    zero,  8
+    vld          vr0,   a1,    0
+    vldx         vr1,   a1,    a5
+    vld          vr2,   a3,    0
+    vld          vr13,  a0,    0
+    vld          vr14,  a0,    16
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vslli.b      vr9,   vr8,   2
+    blt          t1,    t4,    .resS_X_e2
+    fst.d        f6,    a4,    1
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    addi.d       t1,    t1,    -8
+    beqz         t1,    .endS_X_e2
+    vmov         vr13,  vr14
+    addi.d       a4,    a4,    8
+    vbsrl.v      vr9,   vr9,   8
+    vbsrl.v      vr6,   vr6,   8
+.resS_X_e2:
+    vpickve2gr.b t4,    vr9,   0
+    vpickve2gr.h t8,    vr13,  0
+    ldx.w        t6,    sp,    t4
+    ldx.w        t7,    t5,    t4
+    add.w        t6,    t6,    t8
+    addi.w       t7,    t7,    1
+    stx.w        t6,    sp,    t4
+    stx.w        t7,    t5,    t4
+    vstelm.b     vr6,   a4,    1,   0
+    vbsrl.v      vr9,   vr9,   1
+    vbsrl.v      vr13,  vr13,  2
+    vbsrl.v      vr6,   vr6,   1
+    addi.d       t1,    t1,    -1
+    addi.d       a4,    a4,    1
+    bnez         t1,    .resS_X_e2
+.endS_X_e2:
+    ld.d         a0,    sp,    40
+    ld.d         a1,    sp,    48
+    ld.d         a3,    sp,    56
+    ld.d         a4,    sp,    64
+    addi.d       a6,    a6,    -1
+    addi.d       a0,    a0,    128
+    add.d        a1,    a1,    a2
+    st.d         a0,    sp,    40
+    st.d         a1,    sp,    48
+    st.d         a4,    sp,    56
+    st.d         a3,    sp,    64
+    bnez         a6,    .loopS_Y_e2
+    ld.d         t0,    sp,    72
+    vld          vr1,   sp,    0
+    vld          vr2,   t5,    0
+    vld          vr3,   a7,    0
+    vld          vr4,   t0,    0
+    vshuf4i.w    vr5,   vr1,   0xD2
+    vshuf4i.w    vr6,   vr2,   0xD2
+    vadd.w       vr5,   vr5,   vr3
+    vadd.w       vr6,   vr6,   vr4
+    ld.w         t2,    a7,    16
+    ld.w         t3,    t0,    16
+    ld.w         t4,    sp,    16
+    ld.w         t6,    t5,    16
+    add.w        t2,    t2,    t4
+    add.w        t3,    t3,    t6
+    vst          vr5,   a7,    0
+    vst          vr6,   t0,    0
+    st.w         t2,    a7,    16
+    st.w         t3,    t0,    16
+    addi.d       sp,    sp,    72
+endfunc
+
+/*x265_saoCuStatsE3_lsx(const int16_t *diff, const pixel *rec, intptr_t stride, int8_t *upBuff1,
+                        int endX, int endY, int32_t *stats, int32_t *count)
+*/
+function x265_saoCuStatsE3_lsx
+    addi.d       sp,    sp,    -64
+    srli.d       t2,    a4,    4
+    addi.d       t3,    a2,    -1
+    vldi         vr10,  1
+    vldi         vr11,  0
+    vst          vr11,  sp,    0
+    vst          vr11,  sp,    16
+    vst          vr11,  sp,    32
+    st.d         a0,    sp,    40
+    st.d         a1,    sp,    48
+    st.d         a3,    sp,    56
+    addi.d       t5,    sp,    20
+.loopS_Y_e3:
+    ld.d         a0,    sp,    40
+    ld.d         a1,    sp,    48
+    ld.d         a3,    sp,    56
+    move         t0,    t2
+    andi         t1,    a4,    15
+    beqz         t0,    .ResS_X_e3
+.loopS_X_e3:
+    vld          vr0,   a1,    0
+    vldx         vr1,   a1,    t3
+    vld          vr2,   a3,    0
+    vld          vr13,  a0,    0
+    vld          vr14,  a0,    16
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vst          vr6,   a3,    -1
+    vslli.b      vr9,   vr8,   2
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    STATS_COUNT vr9, vr14, 8,  0
+    STATS_COUNT vr9, vr14, 9,  1
+    STATS_COUNT vr9, vr14, 10, 2
+    STATS_COUNT vr9, vr14, 11, 3
+    STATS_COUNT vr9, vr14, 12, 4
+    STATS_COUNT vr9, vr14, 13, 5
+    STATS_COUNT vr9, vr14, 14, 6
+    STATS_COUNT vr9, vr14, 15, 7
+    addi.d       t0,    t0,    -1
+    addi.d       a1,    a1,    16
+    addi.d       a0,    a0,    32
+    addi.d       a3,    a3,    16
+    bnez         t0,    .loopS_X_e3
+.ResS_X_e3:
+    beqz         t1,    .endS_X_e3
+    addi.d       t4,    zero,  8
+    vld          vr0,   a1,    0
+    vldx         vr1,   a1,    t3
+    vld          vr2,   a3,    0
+    vld          vr13,  a0,    0
+    vld          vr14,  a0,    16
+    add.d        a1,    a1,    t1
+    vslt.bu      vr3,   vr1,   vr0
+    vslt.bu      vr4,   vr0,   vr1
+    vand.v       vr3,   vr3,   vr10
+    vor.v        vr5,   vr3,   vr4
+    vneg.b       vr6,   vr5
+    vadd.b       vr7,   vr5,   vr2
+    vaddi.bu     vr8,   vr7,   2
+    vslli.b      vr9,   vr8,   2
+    blt          t1,    t4,    .resS_X_e3
+    fst.d        f6,    a3,    -1
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT vr9, vr13, \i, \i
+.endr
+    addi.d       t1,    t1,    -8
+    addi.d       a3,    a3,    8
+    beqz         t1,    .endS_X_e3
+    vmov         vr13,  vr14
+    vbsrl.v      vr9,   vr9,   8
+    vbsrl.v      vr6,   vr6,   8
+.resS_X_e3:
+    vpickve2gr.b t4,    vr9,   0
+    vpickve2gr.h t8,    vr13,  0
+    ldx.w        t6,    sp,    t4
+    ldx.w        t7,    t5,    t4
+    add.w        t6,    t6,    t8
+    addi.w       t7,    t7,    1
+    stx.w        t6,    sp,    t4
+    stx.w        t7,    t5,    t4
+    vstelm.b     vr6,   a3,    -1,   0
+    vbsrl.v      vr9,   vr9,   1
+    vbsrl.v      vr13,  vr13,  2
+    vbsrl.v      vr6,   vr6,   1
+    addi.d       t1,    t1,    -1
+    addi.d       a3,    a3,    1
+    bnez         t1,    .resS_X_e3
+.endS_X_e3:
+    ldx.bu       t0,    a1,    t3
+    ld.bu        t1,    a1,    0
+    addi.d       t4,    zero,  -1
+    sltu         t7,    t0,    t1
+    sltu         t6,    t1,    t0
+    maskeqz      t8,    t4,    t7
+    addi.d       a5,    a5,    -1
+    or           t6,    t6,    t8
+    st.b         t6,    a3,    -1
+    ld.d         a0,    sp,    40
+    ld.d         a1,    sp,    48
+    addi.d       a0,    a0,    128
+    add.d        a1,    a1,    a2
+    st.d         a0,    sp,    40
+    st.d         a1,    sp,    48
+    bnez         a5,    .loopS_Y_e3
+    vld          vr1,   sp,    0
+    vld          vr2,   t5,    0
+    vld          vr3,   a6,    0
+    vld          vr4,   a7,    0
+    vshuf4i.w    vr5,   vr1,   0xD2
+    vshuf4i.w    vr6,   vr2,   0xD2
+    vadd.w       vr5,   vr5,   vr3
+    vadd.w       vr6,   vr6,   vr4
+    ld.w         t2,    a6,    16
+    ld.w         t3,    a7,    16
+    ld.w         t4,    sp,    16
+    ld.w         t6,    t5,    16
+    add.w        t2,    t2,    t4
+    add.w        t3,    t3,    t6
+    vst          vr5,   a6,    0
+    vst          vr6,   a7,    0
+    st.w         t2,    a6,    16
+    st.w         t3,    a7,    16
+    addi.d       sp,    sp,    64
+endfunc
+
+.macro STATS_COUNT_B in0, in1, i0, i1
+    vpickve2gr.b  t3,  \in0,   \i0
+    vpickve2gr.h  t6,  \in1,   \i1
+    ldx.w         t7,  a5,     t3
+    ldx.w         t8,  a6,     t3
+    add.w         t7,  t7,     t6
+    addi.w        t8,  t8,     1
+    stx.w         t7,  a5,     t3
+    stx.w         t8,  a6,     t3
+.endm
+/*x265_saoCuStatsB0_lsx(const int16_t *diff, const pixel *rec, intptr_t stride,
+                        int endX, int endY, int32_t *stats, int32_t *count)
+*/
+function x265_saoCuStatsB0_lsx
+    srli.d       t2,    a3,    4
+    addi.d       a7,    zero,  8
+.loopS_Y_b0:
+    move         t0,    t2
+    andi         t1,    a3,    15
+    move         t5,    a1
+    move         t4,    a0
+    beqz         t0,    .ResS_X_b0
+.loopS_X_b0:
+    vld          vr0,   t5,    0
+    vld          vr2,   t4,    0
+    vld          vr3,   t4,    16
+    vsrli.b      vr4,   vr0,   3
+    vslli.b      vr5,   vr4,   2
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT_B vr5, vr2, \i, \i
+.endr
+    STATS_COUNT_B vr5, vr3, 8,  0
+    STATS_COUNT_B vr5, vr3, 9,  1
+    STATS_COUNT_B vr5, vr3, 10, 2
+    STATS_COUNT_B vr5, vr3, 11, 3
+    STATS_COUNT_B vr5, vr3, 12, 4
+    STATS_COUNT_B vr5, vr3, 13, 5
+    STATS_COUNT_B vr5, vr3, 14, 6
+    STATS_COUNT_B vr5, vr3, 15, 7
+    addi.d       t0,    t0,    -1
+    addi.d       t5,    t5,    16
+    addi.d       t4,    t4,    32
+    bnez         t0,    .loopS_X_b0
+.ResS_X_b0:
+    beqz         t1,    .endS_X_b0
+    vld          vr0,   t5,    0
+    vld          vr2,   t4,    0
+    vld          vr3,   t4,    16
+    vsrli.b      vr4,   vr0,   3
+    vslli.b      vr5,   vr4,   2
+    blt          t1,    a7,    .resS_X_b0
+.irp i, 0, 1, 2, 3, 4, 5, 6, 7
+    STATS_COUNT_B vr5, vr2, \i, \i
+.endr
+    addi.d       t1,    t1,    -8
+    beqz         t1,    .endS_X_b0
+    vmov         vr2,   vr3
+    vbsrl.v      vr5,   vr5,   8
+.resS_X_b0:
+    STATS_COUNT_B vr5, vr2,  0,  0
+    vbsrl.v      vr5,   vr5,   1
+    vbsrl.v      vr2,   vr2,   2
+    addi.d       t1,    t1,    -1
+    bnez         t1,    .resS_X_b0
+.endS_X_b0:
+    addi.d       a4,    a4,    -1
+    add.d        a1,    a1,    a2
+    addi.d       a0,    a0,    128
+    bnez         a4,    .loopS_Y_b0
+endfunc
diff --git a/source/common/loongarch64/loopfilter.h b/source/common/loongarch64/loopfilter.h
new file mode 100644
index 000000000..85534bb5d
--- /dev/null
+++ b/source/common/loongarch64/loopfilter.h
@@ -0,0 +1,57 @@
+/*****************************************************************************
+ * Copyright (C) 2013-2024 MulticoreWare, Inc
+ *
+ * Authors: Hao Chen <chenhao@loongson.cn>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02111, USA.
+ *
+ * This program is also available under a commercial proprietary license.
+ * For more information, contact us at license @ x265.com.
+ *****************************************************************************/
+
+#ifndef X265_LOOPFILTER_H
+#define X265_LOOPFILTER_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#define DECL_SAO(cpu) \
+    void PFX(saoCuOrgE0_ ## cpu)(pixel * rec, int8_t * offsetEo, int endX, int8_t* signLeft, intptr_t stride); \
+    void PFX(saoCuOrgE1_ ## cpu)(pixel* rec, int8_t* upBuff1, int8_t* offsetEo, intptr_t stride, int width); \
+    void PFX(saoCuOrgE1_2Rows_ ## cpu)(pixel* rec, int8_t* upBuff1, int8_t* offsetEo, intptr_t stride, int width); \
+    void PFX(saoCuOrgE2_ ## cpu)(pixel* rec, int8_t* pBufft, int8_t* pBuff1, int8_t* offsetEo, int lcuWidth, intptr_t stride); \
+    void PFX(saoCuOrgE3_ ## cpu)(pixel *rec, int8_t *upBuff1, int8_t *m_offsetEo, intptr_t stride, int startX, int endX); \
+    void PFX(saoCuOrgB0_ ## cpu)(pixel* rec, const int8_t* offsetBo, int ctuWidth, int ctuHeight, intptr_t stride); \
+    void PFX(calSign_ ## cpu)(int8_t *dst, const pixel *src1, const pixel *src2, const int endX); \
+    void PFX(saoCuStatsE0_ ## cpu)(const int16_t *diff, const pixel * rec, intptr_t stride,  int endX, int endY, int32_t *stats, int32_t *count); \
+    void PFX(saoCuStatsE1_ ## cpu)(const int16_t *diff, const pixel *rec, intptr_t stride, int8_t *upBuff1, int endX, int endY, int32_t *stats, int32_t *count); \
+    void PFX(saoCuStatsE2_ ## cpu)(const int16_t *diff, const pixel *rec, intptr_t stride, int8_t *upBuff1, int8_t *upBufft, int endX, int endY, int32_t *stats, int32_t *count); \
+    void PFX(saoCuStatsE3_ ## cpu)(const int16_t *diff, const pixel *rec, intptr_t stride, int8_t *upBuff1, int endX, int endY, int32_t *stats, int32_t *count); \
+    void PFX(saoCuStatsB0_ ## cpu)(const int16_t *diff, const pixel *rec, intptr_t stride, int endX, int endY, int32_t *stats, int32_t *count); \
+
+DECL_SAO(lsx);
+
+void PFX(calSign_lasx)(int8_t *dst, const pixel *src1, const pixel *src2, const int endX);
+void PFX(pelFilterLumaStrong_V_lsx)(pixel* src, intptr_t srcStep, intptr_t offset, int32_t tcP, int32_t tcQ);
+void PFX(pelFilterLumaStrong_H_lsx)(pixel* src, intptr_t srcStep, intptr_t offset, int32_t tcP, int32_t tcQ);
+void PFX(pelFilterChroma_V_lsx)(pixel* src, intptr_t srcStep, intptr_t offset, int32_t tc, int32_t maskP, int32_t maskQ);
+void PFX(pelFilterChroma_H_lsx)(pixel* src, intptr_t srcStep, intptr_t offset, int32_t tc, int32_t maskP, int32_t maskQ);
+
+#ifdef __cplusplus
+}
+#endif /*__cplusplus*/
+
+#endif // ifndef X265_LOOPFILTER_H
diff --git a/source/test/testharness.h b/source/test/testharness.h
index 5b8e274b1..9342ac416 100644
--- a/source/test/testharness.h
+++ b/source/test/testharness.h
@@ -89,6 +89,9 @@ static inline uint32_t __rdtsc(void)
     a = clock();
 #elif  X265_ARCH_ARM64
     asm volatile("mrs %0, cntvct_el0" : "=r"(a));
+#elif  X265_ARCH_LOONGARCH64
+    uint32_t id = 0;
+    asm volatile("rdtime.d  %0,  %1": "=r"(a), "=r"(id));
 #endif
     return a;
 }
@@ -138,6 +141,8 @@ int PFX(stack_pagealign)(int (*func)(), int align);
  * needs an explicit asm check because it only sometimes crashes in normal use. */
 intptr_t PFX(checkasm_call)(intptr_t (*func)(), int *ok, ...);
 float PFX(checkasm_call_float)(float (*func)(), int *ok, ...);
+#elif (X265_ARCH_LOONGARCH64)
+int PFX(stack_pagealign)(int (*func)(), int align);
 #elif (X265_ARCH_ARM == 0 && X265_ARCH_ARM64 == 0)
 #define PFX(stack_pagealign)(func, align) func()
 #endif
-- 
2.52.0

